File: server/app/suno_internal_output_strip.py
"""Strip internal cognition blocks leaked by the model (e.g. master_blueprint)."""

from __future__ import annotations

import re

_INTERNAL_TAGS = ("master_blueprint", "psychology_audit", "lyric_audit")


def _strip_tag_block(text: str, tag: str) -> str:
    closed = re.compile(
        rf"<{tag}\b[^>]*>.*?</{tag}>",
        re.IGNORECASE | re.DOTALL,
    )
    open_only = re.compile(
        rf"<{tag}\b[^>]*>.*",
        re.IGNORECASE | re.DOTALL,
    )
    out = closed.sub("", text)
    if f"<{tag}" in out.lower():
        out = open_only.sub("", out)
    return out


def strip_internal_cognition_blocks(text: str) -> str:
    """Remove hidden chain-of-thought blocks; keep BLOCK 1 + BLOCK 2 product shape."""
    if not text:
        return text
    lower = text.lower()
    if not any(f"<{tag}" in lower for tag in _INTERNAL_TAGS):
        return text
    out = text
    for tag in _INTERNAL_TAGS:
        out = _strip_tag_block(out, tag)
    out = re.sub(r"\n{3,}", "\n\n", out)
    return out.strip()

File: lib/core/utils/suno_internal_output_strip.dart
/// Internal cognition tags stripped before user-visible Suno output.
const _internalCognitionTags = [
  'master_blueprint',
  'psychology_audit',
  'lyric_audit',
];

/// Strip internal cognition blocks leaked by the model (e.g. master_blueprint).
String stripInternalCognitionBlocks(String text) {
  if (text.isEmpty) return text;
  final lower = text.toLowerCase();
  if (!_internalCognitionTags.any((tag) => lower.contains('<$tag'))) {
    return text;
  }

  var out = text;
  for (final tag in _internalCognitionTags) {
    out = out.replaceAll(
      RegExp(
        '<$tag\\b[^>]*>.*?</$tag>',
        caseSensitive: false,
        dotAll: true,
      ),
      '',
    );
    if (out.toLowerCase().contains('<$tag')) {
      out = out.replaceAll(
        RegExp(
          '<$tag\\b[^>]*>.*',
          caseSensitive: false,
          dotAll: true,
        ),
        '',
      );
    }
  }
  out = out.replaceAll(RegExp(r'\n{3,}'), '\n\n');
  return out.trim();
}
